sphinx/builders/gettext.py

```diff
@@ -57,7 +57,7 @@ def add(self, msg: str, origin: Union[Element, "MsgOrigin"]) -> None:
 
     def __iter__(self) -> Generator[Message, None, None]:
         for message in self.messages:
-            positions = [(source, line) for source, line, uuid in self.metadata[message]]
+            positions = [(os.path.relpath(source, start=os.getcwd()), line) for source, line, uuid in self.metadata[message]]
```

Comment:
This would call `os.getcwd()` for every item in the loop. Does ommitting `start=` still give a usable result? Otherwise you could use cache `os.getcwd()` in a local.

Reply:
Understandable change.

But this should be moved to the outside of this class because this class does not have appropriate directory information. I think what directory you want is the base directory of the document (a.k.a srcdir). But it seems you're using `os.getcwd()`. But it does not equal to srcdir. It usually equals to the confdir that contains conf.py, but not promised. It sometimes returns a different directory if 3rd party extension changes it.


BTW, please use `sphinx.util.osutil:relpath()` instead of `os.path.relpath()`. It will not be crashed even if `path` and `start` are on different drives on Windows.

Reply:
Fixed the issue #10104 after the code reviews from Adam Turner and Takeshi Komiya:
- Added the class variable 'builder' to Catalog class, so any instances of Catalog can access the information within the builder, including things such as app, config, srcdir, outdir etc.. see MessageCatalogBuilder for details.
- Initialise this Catalog.builder as the instance of MessageCatalogBuilder is created, ie. in the init()
- As Message class is created, list of locations passing to it will be made singular, using set(), sorted alphabetically for easier searching and identifying origins where the message can be found.
- source dir is used within the __iter__ of the Catalog, using sphinx.util.osutil.relpath, for safer functionality, to work out the relative path for each location entry.

Reply:
> Added the class variable 'builder' to Catalog class

Modifying global state like this is a bad idea, please could you refactor?

A

Reply:
Yes, adding a class variable is not a good idea. `Catalog` class is a data class that represents message catalog. It is not related to the builder at all. So the path manipulation using builder should be called outside of the catalog class. How about processing just before calling `GettextRenderer`?